src/core/sentinel.d

```diff
@@ -0,0 +1,385 @@
+/**
+Contains types to differentiate arrays with sentinel values.
+*/
+module core.sentinel;
+
+/**
+Selects the default sentinel value for a type `T`.
+
+It has a special case for the char types, and also allows
+the type to define its own default sentinel value if it
+has the member `defaultSentinel`. Otherwise, it uses `T.init`.
+*/
+private template defaultSentinel(T)
+{
+    import std.traits : hasMember, Unqual;
+         static if (is(Unqual!T ==  char)) enum defaultSentinel = '\0';
+    else static if (is(Unqual!T == wchar)) enum defaultSentinel = cast(wchar)'\0';
+    else static if (is(Unqual!T == dchar)) enum defaultSentinel = cast(dchar)'\0';
+    else static if (hasMember(T, "defaultSentinel")) enum defaultSentinel = T.defaultSentinel;
+    else                                   enum defaultSentinel = T.init;
+}
+
+// NOTE: T should be unqalified (not const/immutable etc)
+private template SentinelTemplate(T, immutable T sentinelValue)
+{
+    private enum CommonPtrMembers = q{
+        /**
+        Converts the ptr to an array by scanning it for the sentinel value to find its length.
+
+        Returns:
+            the ptr as a SentinelArray
+        */
+        inout(ArrayType) toArray() inout
+        {
+            ArrayType array = void;
+            array.array = cast(typeof(array.array))ptr[0 .. findLength()];
+            return cast(inout(ArrayType))array;
+        }
+    };
+    struct MutablePtr
+    {
+        alias ArrayType = MutableArray;
+
+        union
+        {
+            T* ptr;
+            ConstPtr asConst;
+        }
+        alias asConst this; // facilitates implicit conversion to const type
+        // alias ptr this; // NEED MULTIPLE ALIAS THIS!!!
+        mixin(CommonPtrMembers);
+
+        /**
+        Coerce the given `array` to a `SentinelPtr`. It checks and asserts
+        if the given array does not contain the sentinel value at `array.ptr[array.length]`.
+        */
+        this(T[] array)
+        in { assert(array.ptr[array.length] == sentinelValue,
```

Comment:
1) Using a contract/assert means this is only enforced in debug builds. This does not improve the safety of the program in terms of the `@safe` attribute.
2) You're going out of bounds here. You can't assume that `array.ptr[array.length]` is supposed to be a sentinel, even if it happens to have the right value.

Same throughout.

Reply:
> 1. Using a contract/assert means this is only enforced in debug builds. This does not improve the safety of the program in terms of the @safe attribute.

I'm not sure, you could either go with `assert` or `enforce`.  It's synonymous with array bounds checking, you may only want it turned on in `debug` mode.  Maybe this check should be "tied into" the same setting, only enabled when array bounds checking is enabled.

> 2. You're going out of bounds here. You can't assume that array.ptr[array.length] is supposed to be a sentinel, even if it happens to have the right value.

Right that's the point of this type.  You can't go outside of the bounds of an array (or a "slice") unless you know the array also owns the elements you are checking.  The point of `SentinalArray` is that if you have one, you know it also "owns" the sentinel element at `array.ptr[array.length]`.  Now this particular constructor is "coercing" a normal array to a `SentinalArray`, and whenever you convert from one type to another, it is up to the developer to make sure the conversion is valid, there's nothing the compiler can do about that.  However, if `SentinalArray` is builtin, meaning that string literals will already be of type `SentinalArray`, then this coercion will only be used as much as normal casting.

Reply:
> It's synonymous with array bounds checking, you may only want it turned on in debug mode.

Note that `@safe` needs bounds checking to work, which is why `-release` keeps bounds checking on in `@safe` code.

> The point of SentinalArray is that you are explicitly declaring it "owns" the sentinel element at array.ptr[array.length].
[...]
> it is up to the developer to make sure the conversion is valid

So getting anything `@safe` is not the plan here? It's just about adding sanity checks to `@system` code?

How about expecting the sentinel in `arr[$ - 1]` (or maybe anywhere in `arr`)? With that and with a run-time check that's always there, you could maybe make it impossible to construct an invalid `SentinelArray`. Then it could be `@safe`:

```d
extern(C) cstring getenv(cstring name) @safe;
extern(C) int puts(cstring s) @safe;
    /* No idea if the char*s are really the only thing keeping these functions from being @safe. */

void main() @safe
{
    cstring p = getenv("PATH\0".asSentinelPtr);
    puts(p);
}
```

(And if `"PATH"` were already a `SentinelArray` as you're planning, it wouldn't need the ugly explicit terminator.)

Reply:
> So getting anything @safe is not the plan here? It's just about adding sanity checks to @system code?

the grand plan would be to change the string literal type to `SentinalString` so you wouldn't have to "coerce" them.  So it should all work in `@safe` code.  The only code that would not be safe would be converting normal D arrays to `SentinalArray` since there's no way to know if an array owns the value `array[$]`.  Actually I also added the `StringLiteral` template which is `@trusted` so you could also use that in `@safe` code, i.e.
```D
puts(StringLiteral!"hello"); // safe
```

> How about expecting the sentinel in arr[$ - 1] (or maybe anywhere in arr)? With that and with a run-time check that's always there, you could maybe make it impossible to construct an invalid SentinelArray. Then it could be @safe:

Good idea.  This is an additional use case that could be used in `@safe`.  I'll see if I can find a good way to add this in.

I know I already answered this, but your getenv example could be written as:
```D
    cstring p = getenv(StringLiteral!"PATH");
```

Reply:
Ok here it is:
```D
/**
This function takes an array whose last element MUST BE the sentinel value and
returns it as a `SentinelArray`.  This differs from the `asSentinelArray` function
which requires not only that the element at `array[$]` is the sentinel value, but also
that the array "owns" the value as well.  This function only operates within the bounds
of the given array, so no "ownership" outside of the array is required.
*/
@property auto reduceToSentinelArray(T)(T[] array) @trusted
in {
    assert(array.length > 0);
    assert(array[$ - 1] == defaultSentinel!T);
   } do
{
    return asSentinelArrayUnchecked(array[0 .. $-1]);
}

///
@safe unittest
{
    auto s = "abc\0".reduceToSentinelArray;
}
```